File: wallet.py
class Portefeuille:
    def __init__(self, solde_initial=0):
        self.solde = solde_initial
        self.historique_transactions = []

    def afficher_solde(self):
        print(f"Solde actuel: {self.solde} euros")

    def deposer(self, montant):
        self.solde += montant
        self.historique_transactions.append(f"Dépôt de {montant} euros")
        print(f"Dépôt de {montant} euros effectué avec succès.")
        self.afficher_solde()

    def retirer(self, montant):
        if montant <= self.solde:
            self.solde -= montant
            self.historique_transactions.append(f"Retrait de {montant} euros")
            print(f"Retrait de {montant} euros effectué avec succès.")
            self.afficher_solde()
        else:
            print("Solde insuffisant pour effectuer le retrait.")

    def afficher_historique(self):
        print("Historique des transactions:")
        for transaction in self.historique_transactions:
            print(transaction)

    def effectuer_virement(self, destinataire, montant):
        if montant <= self.solde:
            self.solde -= montant
            destinataire.deposer(montant)
            self.historique_transactions.append(f"Virement de {montant} euros vers {type(destinataire).__name__}")
            print(f"Virement de {montant} euros vers {type(destinataire).__name__} effectué avec succès.")
            self.afficher_solde()
        else:
            print("Solde insuffisant pour effectuer le virement.")

# Fonction pour exécuter des commandes depuis la console
def executer_commandes():
    portefeuille = Portefeuille()

    while True:
        commande = input("Entrez une commande (solde, depot, retrait, virement, historique, quitter): ").lower()

        if commande == "solde":
            portefeuille.afficher_solde()
        elif commande == "depot":
            montant_depot = float(input("Entrez le montant du dépôt : "))
            portefeuille.deposer(montant_depot)
        elif commande == "retrait":
            montant_retrait = float(input("Entrez le montant du retrait : "))
            portefeuille.retirer(montant_retrait)
        elif commande == "virement":
            destinataire = Portefeuille()
            montant_virement = float(input("Entrez le montant du virement : "))
            portefeuille.effectuer_virement(destinataire, montant_virement)
        elif commande == "historique":
            portefeuille.afficher_historique()
        elif commande == "quitter":
            print("Fermeture du programme.")
            return
        else:
            print("Commande invalide. Veuillez réessayer.")

# Exécutez la fonction pour démarrer le programme interactif
executer_commandes()

File: wallet.js
const readline = require('readline');

class Portefeuille {
    constructor(solde_initial = 0) {
        this.solde = solde_initial;
        this.historique_transactions = [];
    }

    afficherSolde() {
        console.log(`Solde actuel: ${this.solde} euros`);
    }

    deposer(montant) {
        this.solde += montant;
        this.historique_transactions.push(`Dépôt de ${montant} euros`);
        console.log(`Dépôt de ${montant} euros effectué avec succès.`);
        this.afficherSolde();
    }

    retirer(montant) {
        if (montant <= this.solde) {
            this.solde -= montant;
            this.historique_transactions.push(`Retrait de ${montant} euros`);
            console.log(`Retrait de ${montant} euros effectué avec succès.`);
            this.afficherSolde();
        } else {
            console.log("Solde insuffisant pour effectuer le retrait.");
        }
    }

    afficherHistorique() {
        console.log("Historique des transactions:");
        for (const transaction of this.historique_transactions) {
            console.log(transaction);
        }
    }

    effectuerVirement(destinataire, montant) {
        if (montant <= this.solde) {
            this.solde -= montant;
            destinataire.deposer(montant);
            this.historique_transactions.push(`Virement de ${montant} euros vers ${destinataire.constructor.name}`);
            console.log(`Virement de ${montant} euros vers ${destinataire.constructor.name} effectué avec succès.`);
            this.afficherSolde();
        } else {
            console.log("Solde insuffisant pour effectuer le virement.");
        }
    }
}

function executerCommandes() {
    const portefeuille = new Portefeuille();

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    function questionUtilisateur(question) {
        return new Promise(resolve => rl.question(question, resolve));
    }

    async function gestionCommandes() {
        while (true) {
            const commande = (await questionUtilisateur("Entrez une commande (solde, depot, retrait, virement, historique, quitter): ")).toLowerCase();

            if (commande === "solde") {
                portefeuille.afficherSolde();
            } else if (commande === "depot") {
                const montantDepot = parseFloat(await questionUtilisateur("Entrez le montant du dépôt : "));
                portefeuille.deposer(montantDepot);
            } else if (commande === "retrait") {
                const montantRetrait = parseFloat(await questionUtilisateur("Entrez le montant du retrait : "));
                portefeuille.retirer(montantRetrait);
            } else if (commande === "virement") {
                const destinataire = new Portefeuille();
                const montantVirement = parseFloat(await questionUtilisateur("Entrez le montant du virement : "));
                portefeuille.effectuerVirement(destinataire, montantVirement);
            } else if (commande === "historique") {
                portefeuille.afficherHistorique();
            } else if (commande === "quitter") {
                console.log("Fermeture du programme.");
                rl.close();
                return;
            } else {
                console.log("Commande invalide. Veuillez réessayer.");
            }
        }
    }

    gestionCommandes().then(() => rl.close());
}

executerCommandes();
